index.d.ts

```diff
@@ -0,0 +1,156 @@
+declare module 'solc' {
+  export type Primitive =
+    'bool' |
+    'string' |
+    'address' |
+    'uint8' |
+    'uint16' |
+    'uint32' |
+    'uint64' |
+    'uint128' |
+    'uint256' |
+    'int8' |
+    'int16' |
+    'int32' |
+    'int64' |
+    'int128' |
+    'int256' |
+    'bytes' |
+    'bytes20' |
+    'bytes32' |
+    'bool[]' |
+    'string[]' |
+    'address[]' |
+    'uint8[]' |
+    'uint16[]' |
+    'uint32[]' |
+    'uint64[]' |
+    'uint128[]' |
+    'uint256[]' |
+    'int8[]' |
+    'int16[]' |
+    'int32[]' |
+    'int64[]' |
+    'int128[]' |
+    'int256[]' |
+    'bytes[]' |
+    'bytes20[]' |
+    'bytes32[]';
+
+  export interface AbiParameter {
+    name: string,
+    type: Primitive,
+  }
+  
+  export interface AbiEventParameter extends AbiParameter {
+    indexed: boolean,
+  }
+  
+  export interface AbiFunction {
+    name: string,
+    type: 'function' | 'constructor' | 'fallback',
+    stateMutability: 'pure' | 'view' | 'payable' | 'nonpayable',
+    constant: boolean,
+    payable: boolean,
+    inputs: Array<AbiParameter>,
+    outputs: Array<AbiParameter>,
+  }
+  
+  export interface AbiEvent {
+    name: string,
+    type: 'event',
+    inputs: Array<AbiEventParameter>,
+    anonymous: boolean,
+  }
+  
+  export type Abi = Array<AbiFunction | AbiEvent>;
+  
+  interface CompilerInputSourceFile {
+    keccak256?: string;
+    urls: string[];
+  }
+  interface CompilerInputSourceCode {
+    keccak256?: string;
+    content: string;
+  }
+  interface CompilerInput {
+    language: "Solidity" | "serpent" | "lll" | "assembly";
+    settings?: any,
+    sources: {
+      [globalName: string]: CompilerInputSourceFile|CompilerInputSourceCode,
+    };
+  }
+  interface CompilerOutputError {
+    sourceLocation?: {
+      file: string;
+      start: number;
+      end: number;
+    };
+    type: "TypeError" | "InternalCompilerError" | "Exception";
+    component: "general" | "ewasm";
+    severity: "error" | "warning";
+    message: string;
+    formattedMessage?: string;
+  }
+  interface CompilerOutputEvmBytecode {
+    object: string;
+    opcodes?: string;
+    sourceMap?: string;
+    linkReferences?: {} | {
+      [globalName: string]: {
+        [name: string]: {start: number, length: number}[];
+      };
+    };
+  }
+  interface CompilerOutputSources {
+    [globalName: string]: {
+      id: number;
+      ast: any;
+      legacyAST: any;
+    },
+  }
+  interface CompilerOutputContracts {
+    [globalName: string]: {
+      [contractName: string]: {
+        abi: Abi;
```

Comment:
All of these are optional. I'm not sure it is a good idea to have forced description the JSON, it is not actually part of solc-js, but part of Solidity.

Reply:
You mean `abi` and `evm`?  The rest are already marked as optional.

Reply:
Everything in the JSON output is optional as it depends on the JSON input.

Reply:
Is there anywhere that this is documented?  When you say "everything is optional" I'm pretty sure you don't actually mean everything as (for example) I don't believe that you can configure the output to omit `type` in the `AbiFunction` structure.  Perhaps you just mean "each of the top level keys under `[contractName: string]` are optional"?  Are all of them actually optional or are some of them optional in the sense that they may be empty objects or empty arrays?  Looking at the docs, the example shows several as having empty arrays or empty objects and I _vaguely_ remember when I was working with these a while back encountering some situations where rather than being omitted an empty object or array was included (but I could be misremembering).

My goal here is to accurately represent the structure of the object returned by the Solidity compiler.  My inquiries are because I really have no idea exactly what it _can_ spit out, so I need some help specifying it.

If you really want to go above and beyond to help out, a comment on each non-optional thing that can be optional would be ideal.  😄

Reply:
Looking over the docs again, I have added a few more optional indicators (locally).  However, it is unclear from the docs whether `evm.bytecode` will be included if all of the members of `evm.bytecode` are excluded?  Or perhaps it will just be an empty object in this case?  Similar question for `ewasm` section, is it an empty object if `ewasm.wast` and `ewasm.wasm` are filtered out or is it omitted entirely if they are filtered out?

Reply:
Same question for the top level things, will the `contracts` key be omitted entirely if everything is filtered out from it, or will it always be present with an empty object?

Reply:
All of it is documented here: http://solidity.readthedocs.io/en/develop/using-the-compiler.html#compiler-input-and-output-json-description

The ABI is documented here (but that is again outside the scope of the JSON compiler interface, it is a specification shared between languages and clients): http://solidity.readthedocs.io/en/develop/abi-spec.html#json